### DISCUSSION TITLE
Interdependence of some projects in CFU

### DISCUSSION THREAD

**Discussion by Jimrnelson** [VFPX/Thor#121] [General]
Some projects available form CFU assume that other projects have been installed and will not work properly (or maybe not at all) if the other projects are not installed.

Of course, there is no way for users to be aware of this interdependence, so I am opening this discussion about what to do about it.

This issue has come to mind because of a recent issue in Bin2Text, where the user followed the documentation to download Thor and then run CFU to download Bin2Text.  Even though projects "Thor Repository" and "PEM Editor" are checked off, by default, in CFU, the user un-checked them and only installed Bin2Text.  Things failed because Bin2Text (unknown possibly even to the author) depended on "Thor Repository" being installed.  (Installing it took care of at least part of the user's problem).

Thus, it seems to me that at least two projects, "Thor Repository" and "PEM Editor", should _always_ be installed, to avoid instances like this.

After further thought, I want to add "Dynamic Forms" to the list, as tools in "Thor Repository" (such as Super Browse) use Dynamic Forms, as well as some planned enhancements to Thor itself.

A secondary list is the pair of "GoToDefinition" and "IntellisenseX", as there is shared code between them -- neither of them are fully functional without the other.  The change here would be that requesting one would in effect be requesting the other.

I wanted to get feedback on these ideas before beginning implementation.

**Comment by DougHennig** [comment]
Perhaps there could be a Dependencies property of the updater object. It would be set to a list of the AppID of the projects that are also installed if not already. e.g. for Thor, it would be set to the AppID for Thor Repository and PEM Editor. When a project is being installed, CFU checks Dependencies to see if the specified projects are installed and if not, download and install them too (it could just run the Thor_Update_AppID.prg).

**Comment by Jimrnelson** [comment]
RE:  New property "Dependencies"

Super idea.  That will cover all needs, for sure, and at first blush looks like that will be easy to implement.

**Comment by Jimrnelson** [comment]
I have added this new property "Dependencies".

All projects are presumed dependent on PEM Editor, Thor Repository, and Dynamic Forms.

GoToDefinition is dependent on IntellisenseX and uses this new property.

**Reply by lscheffler** [reply]
Do you have a documentation?

**Reply by Jimrnelson** [reply]
What are you asking for here?

**Reply by lscheffler** [reply]
How does this property "Dependencies" work. 
- where to place
- how to enumerate dependencies

---
If possible, can you alter VFPXDeployment and add the property to the templates, with comments?

**Comment by lscheffler** [comment]
I see that "Thor Repository" and "PEM Editor" is installed by default right now (I had some headache about it until I found this discussion).   
The main tools must be available, so, yes.
I see some problems since Thor Repository holds stuff that depends on other projects. For example there is stuff for FoxBin2Prg in it.

**Reply by Jimrnelson** [reply]
@lscheffler @DougHennig 

> I see some problems since Thor Repository holds stuff that depends on other projects. For example there is stuff for FoxBin2Prg in it.

I believe that the idea here is to declare **implicit** dependencies in a project.  That is, dependencies on another project that a user would not necessarily know about.

Unless I am misunderstanding what's going on with FoxBin2PRG, I would not think of it as an implicit dependency in Thor Repository, since the tools that need it explicitly reference FoxBin2PRG.  One could easily argue that those tools should have been part of the FoxBin2PRG project and not Thor Repository, but it's a little late for that.

From my perspective, I believe this discussion is complete as it has been implemented.  If there are any other dependencies identified, the structure is already there to handle them.

**Reply by lscheffler** [reply]
Oh! That was understood by the discussion below. So yes, this is done.

**Comment by Jimrnelson** [comment]
"Dependencies" is a property in the Updater object.  It can be set either in the Update prg for a project or in the version file.

It indicates what other VFPX projects this project requires.

It is a list (comma or space delimited) of the App names of the projects.

So far, only used to link IntellisenseX and GoToDefinition.

**Reply by lscheffler** [reply]
So it can be used in the "declaration" file in Thor as well as in the project's version file?
Bin2Text is a candidate, somehow. :D

**Comment by Jimrnelson** [comment]
I'm not sure what you are referring to when you say a "declaration" file.

Each VFPX project that appears in CFU has an Updater file, managed in Thor.  These are intended to be static.  This is the only file for 3/4 of al projects.

Within the Updater there can be a link to the Version file, which is managed by the individual project.

Properties can be set in either place.  In fact, they can be set in the Updater but then over-ridden by the Version file.

**Reply by lscheffler** [reply]
The static "Updater file" is declaring something to be included into Thor. Missed the word. :)   
Thank you.
